src/main/java/com/google/cloud/tools/app/module/DeleteAction.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools.app.module;
+
+import com.google.cloud.tools.app.Action;
+import com.google.cloud.tools.app.GCloudExecutionException;
+import com.google.cloud.tools.app.Option;
+import com.google.cloud.tools.app.ProcessCaller;
+import com.google.cloud.tools.app.ProcessCaller.ProcessCallerFactory;
+import com.google.cloud.tools.app.ProcessCaller.Tool;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Deletes a version of one or more modules.
+ */
+public class DeleteAction extends Action {
+
+  private static ProcessCallerFactory processCallerFactory = ProcessCaller.getFactory();
+  private Collection<String> modules = UNSET_COLLECTION;
```

Comment:
It's pretty odd to do this with null.  In your example they have a sentinal
value  for ints (cause you can't assign null to a primitive), and stub
implementations for null objects.

I'm not sure where getting much mileage using this pattern for String,
Integer, etc.  Also I don't think it's worth doing the precondition checks
that they aren't setting the same value twice.  It's actually a reasonable
thing to do, but not worth the code IMO.  I do think ti would be worthwhile
to check our arguments to make sure they aren't null or some other invalid
value.

On Wed, Mar 23, 2016 at 1:36 PM, João Martins <notifications@github.com>
wrote:

> In src/main/java/com/google/cloud/tools/app/module/DeleteAction.java
> <https://github.com/GoogleCloudPlatform/gcp-tools-lib-for-java/pull/1#discussion_r57203059>
> :
>
> > +import com.google.cloud.tools.app.ProcessCaller.Tool;
> > +import com.google.common.annotations.VisibleForTesting;
> > +import com.google.common.base.Preconditions;
> > +import com.google.common.base.Strings;
> > +
> > +import java.util.ArrayList;
> > +import java.util.Collection;
> > +import java.util.List;
> > +
> > +/**
> > + * Deletes a version of one or more modules.
> > + */
> > +public class DeleteAction extends Action {
> > +
> > +  private static ProcessCallerFactory processCallerFactory = ProcessCaller.getFactory();
> > +  private Collection<String> modules = UNSET_COLLECTION;
>
> I understand that and can live with it. If the final value is null, the
> operation will always fail if the parameter is required.
>
> As I said, this is to improve code readability. It is a best practice
> within Google and I didn't pioneer it.
> There is an example of its usage at
> https://github.com/google/guava/blob/024e56a668b8a1d2653655a5530ed9f4ed5bb166/guava/src/com/google/common/cache/CacheBuilder.java
> (UNSET_INT) to virtualize "final".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcp-tools-lib-for-java/pull/1/files/10d1211864218e1bd38d247da31c02ff9d402984#r57203059>
>